AO3-5610 notify users when a bookmark or series they have created is hidden

walshyb opened the pull request:
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [ ] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-5610 (Please fill in issue number and remove this comment.)

## Purpose

What does this PR do?

## Testing Instructions

How can the Archive's QA team verify that this is working as you intended?

If you have a Jira account with access, please update or comment on the issue
with any new or missing testing instructions instead.

## References

Are there other relevant issues/pull requests/mailing list discussions?

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?

If you have a Jira account, please include the same name in the "Full name"
field on your Jira profile, so we can assign you the issues you're working on.

*Please note that if you do not fill in this section, we will use your GitHub account name and
they/them pronouns.*

walshyb committed:
Create series notification mailer action

walshyb committed:
Create series model admin_hidden update handler

walshyb committed:
Add series mailer content and en translation

walshyb commented on app/mailers/user_mailer.rb:
```diff
@@ -377,6 +377,18 @@ def admin_deleted_work_notification(user, work)
     end
   end
 
+  def admin_hidden_series_notification(creation_id, user_id)
```
do we want to reutilize the `admin_hidden_work_notification` action, and maybe call it `admin_hidden_creation_notification`?

and update the email contents by passing in parameters?

sarken replied:
Personally, I like separate actions for each. I think the contents of the emails could potentially have enough differences that it could get kind of messy trying to account for all of the variations.

walshyb replied:
Thank you! 

Added a separate action for bookmarks

walshyb committed:
remove unnecessary end

walshyb commented:
I haven't used Rails + i18n before. Is there a way to generate the translations from the yaml i added to `config/locales/mailers/en.yml`?

sarken commented:
If you mean a way to, e.g., translate the English text into Spanish and insert it in es.yml, nope! All our translations are done by the wonderful human translators on our Translation team. They enter it in [Phrase](https://phrase.com) and then export it to GitHub when it's ready (pull request #4517 is an example).

Bilka2 commented on app/views/user_mailer/admin_hidden_series_notification.html.erb:
```diff
@@ -0,0 +1,14 @@
+
+  <p><%= t(".check_email") %></p>
+
+  <p><%= t(".html.tos_violation", tos_link: tos_link(t ".tos")).html_safe %></p>
```
The ideal configuration of parentheses would be this:
```suggestion
  <p><%= t(".tos_violation.html", tos_link: tos_link(t(".tos"))) %></p>
```

Similar for the abuse link below.

Bilka2 commented on config/locales/mailers/en.yml:
```diff
@@ -6,20 +6,20 @@ en:
           one: If you do not use this link to reset your password within %{count} day, it will expire, and you will have to request a new one.
           other: If you do not use this link to reset your password within %{count} days, it will expire, and you will have to request a new one.
-        intro: 'Someone has requested a password reset for your account. You can change your account password by following the link below and entering your new password:'
+        intro: "Someone has requested a password reset for your account. You can change your account password by following the link below and entering your new password:"
```
The style guideline regarding `"` quotation marks being preferred does not apply to the locale yml files. The normalized files are expected to use `'` when needed.

You can see this because the i18n test for this file [fails](https://github.com/otwcode/otwarchive/actions/runs/8118300579/job/22192291997?pr=4760#step:12:1592). To have to show up as a failed test here on github, merge `master` into your branch so that you get [this bug fix](https://github.com/otwcode/otwarchive/commit/321fee426a60d87b5cd918fec93ec90a8c367116).

Bilka2 replied:
(You can normalize the file again with `bundle exec i18n-tasks normalize -l en`, from https://github.com/otwcode/otwarchive/wiki/Internationalization-%28i18n%29-Standards#i18n-tasks.)

walshyb replied:
I ran `bundle exec i18n-tasks normalize -l en # Normalize formatting of en.yml` in my docker container and it didn't appear to change the file (in the container, nor in my local codebase). 

I can make the changes manually, though. Just to be clear, all instances of `"` (double quotes) should be `'` (single quotes)?

Thank you

walshyb replied:
Actually maybe that command fixed it because I just checked and most of the cases where `"` was used now it's `'`. So we may be good!

Bilka2 commented on app/mailers/user_mailer.rb:
```diff
@@ -377,6 +377,18 @@ def admin_deleted_work_notification(user, work)
+    I18n.with_locale(@user.preference.locale.iso) do
+      mail(
+        to: @user.email,
+        subject: default_i18n_subject(app_name: ArchiveConfig.APP_SHORT_NAME)
```
This needs a comment to help out i18n-tasks:
```suggestion
        # i18n-tasks-use t('user_mailer.admin_hidden_series_notification.subject')
        subject: default_i18n_subject(app_name: ArchiveConfig.APP_SHORT_NAME)
```

Bilka2 replied:
Update: The comment is no longer needed! Just merge the current master branch into this PR branch and i18n-tasks will recognize `default_i18n_subject` automatically.

Bilka2 reviewed (commented):
I left some comments regarding the localisation of the email. There are so many notes because the admin_hidden_work_notification that you based this of was using some outdated I18n standards.

Bilka2 commented:
Noting that besides the i18n review comments, the notification for bookmarks still needs to be added.

walshyb committed:
Add hidden bookmark notification email

walshyb committed:
Remove explicit html_safe call

walshyb committed:
Merge master

walshyb committed:
Add hidden bookmark mailer and preview

walshyb committed:
Update hidden bookmark mailer locale

walshyb committed:
Update hidden series mailer locale and add preview

walshyb committed:
Remove explicit html_safe

walshyb committed:
Fix incorrect filename

walshyb committed:
Remove unnecessary parentheses

walshyb commented:
Hello!

I've made the following changes per the above comments:
- Added an `admin_hidden_series_notification` mailer preview
- Added an `admin_hidden_bookmark_notification` mailer action, lifecycle callback, view, preview, and locale stuff
- Updated the respective locale content to fit the current standard

Questions:
- If possible, would you like me to write explicit tests to ensure the mails are getting sent (or I guess added to the queue) when an admin hides the respective models?
- Can you double check the verbiage used in the new mailers? 
- Did I follow all the i18n standards?

Thank you!

walshyb committed:
Linting fixes

Bilka2 commented on app/models/series.rb:
```diff
@@ -22,6 +22,8 @@ class Series < ApplicationRecord
     maximum: ArchiveConfig.TITLE_MAX,
     too_long: ts("must be less than %{max} letters long.", max: ArchiveConfig.TITLE_MAX)
 
+  after_update :admin_hidden_series_notification, if: :hidden_by_admin_changed?
```
Can you move this to be one line above the `after_save` callback below, please

Bilka2 commented on app/models/bookmark.rb:
```diff
@@ -32,6 +32,8 @@ def check_new_external_work
     end
   end
 
+  after_update :admin_hidden_bookmark_notification, if: :hidden_by_admin_changed?
```
Can you move this to be one (or two) line(s) above the `before_destroy` callback below, please

Bilka2 commented on app/views/user_mailer/admin_hidden_bookmark_notification.html.erb:
```diff
@@ -0,0 +1,13 @@
+<% content_for :message do %>
+  <p><%= t("mailer.general.greeting.formal_html", name: style_bold(@user.login)) %></p>
+
+  <p><%= t(".hidden.html", bookmark_url: style_creation_link("bookmark", @bookmark)) %></p>
```
The link text should be "bookmark of %{title of bookmarkable item}" and should be translatable.

Also, the variable here should be `bookmark_link` instead of `bookmark_url`.

Note also from Jira: In the HTML email, the creation title inside the bookmark link should be bold and italicized.

Bilka2 commented on app/views/user_mailer/admin_hidden_bookmark_notification.text.erb:
```diff
@@ -0,0 +1,13 @@
+<% content_for :message do %>
+<%= t("mailer.general.greeting.formal_html", name: @user.login) %>
+
+<%= t(".hidden.text", bookmark_url: bookmark_url(@bookmark)) %>
```
The text here should be `Your bookmark of "<%= @bookmark.bookmarkable.title %>" (<%= bookmark_url(@bookmark) %>) has ...` so something similar to what you did for the series.

walshyb replied:
I did the same thing that I did for the series mailer. Is this new change good?

Bilka2 replied:
The text is good now, but the variable name is incorrect. I left a comment on the line in the yml file.

Bilka2 commented on config/locales/mailers/en.yml:
```diff
@@ -120,6 +120,36 @@ en:
+        html: If your bookmark was hidden due to being in violation of the Archive of Our Own's %{tos_link}, you will be required to take action to correct the violation. Failure to bring your bookmark into compliance with the Terms of Service may lead to your bookmark being deleted from the Archive.
+        text: If your bookmark was hidden due to being in violation of the Archive of Our Own's Terms of Service (%{tos_url}), you will be required to take action to correct the violation. Failure to bring your bookmark into compliance with the Terms of Service may lead to your bookmark being deleted from the Archive.
+    admin_hidden_series_notification:
+      access: While your series is hidden, you will still be able to access it through the link provided above, but it will not be listed on your series' page, and it won't be available to other users of the Archive.
```
```suggestion
      access: While your series is hidden, you will still be able to access it through the link provided above, but it will not be listed on your series page, and it won't be available to other users of the Archive.
```

Bilka2 commented on test/mailers/previews/user_mailer_preview.rb:
```diff
@@ -40,6 +40,18 @@ def claim_notification_registered
   end
 
+  def admin_hidden_bookmark_notification
+    bookmark = create(:bookmark)
+    creator_id = bookmark.pseud.user_id
+    UserMailer.admin_hidden_bookmark_notification(bookmark.id, creator_id)
```
It'd be nice to have a more "standard" username in the preview:
```suggestion
    creator = create(:user, :for_mailer_preview)
    bookmark = create(:bookmark, pseud: creator.default_pseud)
    UserMailer.admin_hidden_bookmark_notification(bookmark.id, creator.id)
```

Bilka2 commented on test/mailers/previews/user_mailer_preview.rb:
```diff
@@ -40,6 +40,18 @@ def claim_notification_registered
+
+  def admin_hidden_series_notification
+    series = create(:series)
+    creator_id = series.pseuds.first.user_id
+    UserMailer.admin_hidden_series_notification(series.id, creator_id)
```
It'd be nice to have a more "standard" username in the preview:
```suggestion
    creator = create(:user, :for_mailer_preview)
    UserMailer.admin_hidden_series_notification(series.id, creator.id)
```

Bilka2 reviewed (changes requested):
The additions look good, there are just some details and tests left.

> If possible, would you like me to write explicit tests to ensure the mails are getting sent (or I guess added to the queue) when an admin hides the respective models?

That would be nice! That would have to be cucumber tests. It'd be even nicer if you could check that the emails end up getting translated in those tests, by hiding the bookmark/series belonging to two users where one of them has translated emails enabled. An example of a test for translated emails can be seen in #4921.

Furthermore, we usually add rspec tests for new emails, you can take a look at what was done for the admin hidden work notification or in the recent #4952.

One more thing, there is this note on Jira:
> The footer should no longer say, “If you've received this message in error, please contact Support.”

The current email still says this. This footer text is defined in `app/views/layouts/mailer.html.erb` and `app/views/layouts/mailer.text.erb`. It needs to be suppressed for the emails here and `admin_hidden_work_notification`.

I noted the few remaining i18n problems in separate review comments.

walshyb committed:
Merge branch 'master' into AO3-5610-Notify-users-when-a-bookmark-or-series-they-have-created-is-hidden

walshyb committed:
Move placement of after_update callback

walshyb committed:
Move after_update callback

walshyb committed:
Remove unnecessary single quote

Co-authored-by: Bilka <Bilka2@users.noreply.github.com>

walshyb committed:
Create user for hidden bookmark mailer test

Co-authored-by: Bilka <Bilka2@users.noreply.github.com>

walshyb committed:
Create user for hidden series mailer test

Co-authored-by: Bilka <Bilka2@users.noreply.github.com>

walshyb committed:
Change var from bookmark_url to bookmark_link

walshyb committed:
Merge branch 'AO3-5610-Notify-users-when-a-bookmark-or-series-they-have-created-is-hidden' of github.com:walshyb/otwarchive into AO3-5610-Notify-users-when-a-bookmark-or-series-they-have-created-is-hidden

walshyb committed:
Change var from bookmark_url to bookmark_link

walshyb committed:
Update hidden bookmark mailer text to include bookmark title

Bilka2 commented on config/locales/mailers/en.yml:
```diff
@@ -120,6 +120,36 @@ en:
+        html: If you are uncertain why your bookmark was hidden, and you have not received further communication regarding this matter, please %{contact_abuse_link} directly.
+        text: "If you are uncertain why your bookmark was hidden, and you have not received further communication regarding this matter, please contact Policy & Abuse directly: %{contact_abuse_url}."
+      hidden:
+        html: Your bookmark of %{title} has been hidden by the Policy & Abuse team and is no longer publicly accessible.
```
According to Jira, the entire "bookmark of X" text should be the link, so this needs to be split up:
```suggestion
        bookmark_of: bookmark of %{title}
        html: Your %{bookmark_of_link} has been hidden by the Policy & Abuse team and is no longer publicly accessible.
```

For why I named the link variable the way I did, refer to https://github.com/otwcode/otwarchive/wiki/Internationalization-%28i18n%29-Standards#hyperlinks

Bilka2 commented on app/views/user_mailer/admin_hidden_bookmark_notification.html.erb:
```diff
@@ -0,0 +1,13 @@
+<% content_for :message do %>
+  <p><%= t("mailer.general.greeting.formal_html", name: style_bold(@user.login)) %></p>
+
+  <p><%= t(".hidden.html", title: style_creation_link(@bookmark.bookmarkable.title, @bookmark)) %></p>
```
And then to match the change in the yml file and also conform to this from Jira:
> In the HTML email, the creation title inside the bookmark link should be bold and italicized.
 
```suggestion
  <p><%= t(".hidden.html", bookmark_of_link: link_to(t(".hidden.bookmark_of", title: tag.em(tag.strong(@bookmark.bookmarkable.title))), @bookmark)) %></p>
```

Bilka2 commented on config/locales/mailers/en.yml:
```diff
@@ -120,6 +120,36 @@ en:
+        text: "If you are uncertain why your bookmark was hidden, and you have not received further communication regarding this matter, please contact Policy & Abuse directly: %{contact_abuse_url}."
+      hidden:
+        html: Your bookmark of %{title} has been hidden by the Policy & Abuse team and is no longer publicly accessible.
+        text: Your bookmark of "%{title}" (%{bookmark_link}) has been hidden by the Policy & Abuse team and is no longer publicly accessible.
```
In the text version of the email this is a plain URL so the variable name should `bookmark_url`, not _link

(Needs the change in admin_hidden_bookmark_notification.text.erb too)

Bilka2 reviewed (changes requested):
I think an autoformatter decided to change all the quotes in the yml file, please normalize the file again.

To make tracking a bit easier, still missing from my previous review: Cucumber and rspec tests, moving the I18n.with_locale call, removing the notice to contact Support from the footer
